packages/adblocker/src/preprocessor.ts

```diff
@@ -0,0 +1,211 @@
+import { StaticDataView, sizeOfUTF8 } from './data-view';
+import { fastStartsWith } from './utils';
+
+export type EnvKeys =
+  | 'ext_ghostery'
+  | 'ext_abp'
+  | 'ext_ublock'
+  | 'ext_ubol'
+  | 'ext_devbuild'
+  | 'env_chromium'
+  | 'env_edge'
+  | 'env_firefox'
+  | 'env_mobile'
+  | 'env_safari'
+  | 'env_mv3'
+  | 'false'
+  | 'cap_html_filtering'
+  | 'cap_user_stylesheet'
+  | 'adguard'
+  | 'adguard_app_android'
+  | 'adguard_app_ios'
+  | 'adguard_app_mac'
+  | 'adguard_app_windows'
+  | 'adguard_ext_android_cb'
+  | 'adguard_ext_chromium'
+  | 'adguard_ext_edge'
+  | 'adguard_ext_firefox'
+  | 'adguard_ext_opera'
+  | 'adguard_ext_safari'
+  | (string & {});
+
+export class Env extends Map<EnvKeys, boolean> {}
+
+export const enum PreprocessorTypes {
+  INVALID = 0,
+  BEGIF = 1,
+  ELSE = 2,
+  ENDIF = 3,
+}
+
+export function detectPreprocessor(line: string) {
+  if (line.charCodeAt(1) !== 35 /* '#' */) {
+    return PreprocessorTypes.INVALID;
+  }
+
+  const command = line.slice(2);
+
+  if (fastStartsWith(command, 'if ')) {
+    return PreprocessorTypes.BEGIF;
+  }
+
+  if (command === 'else') {
+    return PreprocessorTypes.ELSE;
+  }
+
+  if (command === 'endif') {
+    return PreprocessorTypes.ENDIF;
+  }
+
+  return PreprocessorTypes.INVALID;
+}
+
+const operatorPattern = /(\|\||&&)/g;
+const identifierPattern = /^(!?[a-z0-9_]+)$/;
+
+const tokenize = (expression: string) => expression.split(operatorPattern);
+const matchIdentifier = (identifier: string) => identifier.match(identifierPattern);
+
+export const evaluate = (expression: string, env: Env) => {
```

Comment:
What is this algorithm? :) Does it handle precedence rules between &&, ||, ! as well as grouping with parentheses?

Reply:
The grouping with parentheses was not implemented to follow the specification of uBlock Origin.

Reply:
I am not convinced this new implementation is correct (especially since I did not see any logic to handle precedence of different operators like `&&`, `||`, `!`).

As far as uBlock Original is concerned they say they [support the full syntax](https://adguard.com/kb/general/ad-filtering/create-own-filters/#conditions-directive) which contains parentheses.

Reply:
I implemented a new evaluation function based on your snippet on the Slack. Now it supports parenthesis and negated preprocessors by wrapping the previous condition.

Reply:
@seia-soto lets add a test case for this function.

Reply:
@chrmod I'll try to find more edge cases of `evaluate` function. The test code is already defined in `preprocessor.test.ts`.